File: plugins/player/skill/fishing/fishing_spot.scala
/*
 A plugin for the Fishing skill that adds functionality for moving fishing spots.

 SUPPORTS:
  -> Moving fishing spots around at random intervals.
  -> Two possible positions for a fishing spots to be in.

 TODO:
  -> Moving different fishing spots at different times.

 AUTHOR: lare96
*/

import io.luna.game.event.impl.ServerLaunchEvent
import io.luna.game.model.Position
import io.luna.game.model.mobile.Npc


/* Class representing fishing spots in 'FISHING_SPOTS'. */
private case class FishingSpot(id: Int, homePos: Position, awayPos: Position) {
  val npcInstance = new Npc(ctx, id, homePos)
}


/* A range of how often fishing spots will move, in ticks. */
private val MOVE_INTERVAL = 100 to 2000 // Fishing spots will currently move every 1-20 minutes.

/* A Seq of fishing spots that will periodically move. */
private val FISHING_SPOTS = Seq.empty[FishingSpot]


/* Moves fishing spots from their 'home' to 'away' positions, and vice-versa. */
private def moveFishingSpots = {
  for (spot <- FISHING_SPOTS) {
    val spotNpc = spot.npcInstance

    spotNpc.position match {
      case spot.homePos => spotNpc.teleport(spot.awayPos)
      case spot.awayPos => spotNpc.teleport(spot.homePos)
      case _ => fail("invalid fishing spot position")
    }
  }
}


/* Intercept server launch event, perform startup operations for fishing spots. */
intercept[ServerLaunchEvent] { (msg, plr) =>
  if (FISHING_SPOTS.nonEmpty) {
    FISHING_SPOTS.foreach(spot => world.addNpc(spot.npcInstance))

    world.scheduleInterval(MOVE_INTERVAL) { task =>
      moveFishingSpots
    }
  }
}

File: plugins/world/player/skill/fishing/fishingSpot/FishingSpot.kt
package world.player.skill.fishing.fishingSpot

import api.predef.*
import io.luna.game.model.Position
import io.luna.game.model.mob.Npc

/**
 * A model representing a fishing spot.
 */
class FishingSpot(id: Int,
                  val home: Position,
                  val away: Position) : Npc(ctx, id, home) {

    companion object {

        /**
         * A range of how often fishing spots can move, in minutes.
         */
        val MOVE_INTERVAL = 1..7 // 1-7 minutes.
    }

    /**
     * The countdown timer. This spot will move when it reaches 0.
     */
    var countdown = MOVE_INTERVAL.random()

    /**
     * Performs a countdown and returns true if the spot should be moved.
     */
    fun countdown(): Boolean {
        countdown--
        if (countdown <= 0) {
            countdown = MOVE_INTERVAL.random()
            return true
        }
        return false
    }
}